### PR TITLE
Genric pipe system and manufacturer Product Ranges

### PR DESCRIPTION
**Pull request by RajeevKumarSinha**
In These two I have used Referencing, and Populate to perform CRUD.

### CODE REVIEW

File: controllers/masterPipeController.js [hunk 1]
```diff
@@ -59,13 +59,15 @@ exports.removeMasterPipe = async (req, res, next) => {
 		if (!req.body.ids) return next(errObject("ids not found", 400))
 
 		const idArr = req.body.ids
+		console.log(idArr)
```
Comment (rohandhamapurkar) [inline controllers/masterPipeController.js]: remove console

File: controllers/masterPipeController.js [hunk 2]
```diff
@@ -59,13 +59,15 @@ exports.removeMasterPipe = async (req, res, next) => {
...
 
 		// if masterPipe ids length is 0 throw an error
 		if (idArr.length === 0) {
 			return next(errObject("MasterPipe ids array cannot be empty", 400))
 		}
 
-		await masterPipeService.deleteMasterPipe(idArr)
+		const x = await masterPipeService.deleteMasterPipe(idArr)
+		console.log(x)
```
Comment (rohandhamapurkar) [inline controllers/masterPipeController.js]: remove console

File: controllers/masterPipeController.js [hunk 3]
```diff
@@ -59,13 +59,15 @@ exports.removeMasterPipe = async (req, res, next) => {
...
 		const idArr = req.body.ids
+		console.log(idArr)
 
 		// if masterPipe ids length is 0 throw an error
 		if (idArr.length === 0) {
 			return next(errObject("MasterPipe ids array cannot be empty", 400))
 		}
 
-		await masterPipeService.deleteMasterPipe(idArr)
+		const x = await masterPipeService.deleteMasterPipe(idArr)
```
Comment (rohandhamapurkar) [inline controllers/masterPipeController.js]: unused variable "x"

File: services/genericPipeService.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+			.skip(pageNo * docsPerPage)
+			.limit(docsPerPage)
+
+		const response = {
+			total: totalDocs,
+			genericPipeData,
+		}
+
+		return response
+	} catch (error) {
```
Comment (rohandhamapurkar) [inline services/genericPipeService.js]: this try catch redundant and not needed.
same for all services

File: services/manufacturerProdRangeService.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+			.skip(pageNo * docsPerPage)
+			.limit(docsPerPage)
+
+		const response = {
+			total: totalDocs,
+			genericPipeData,
+		}
+
+		return response
+	} catch (error) {
```
Comment (rohandhamapurkar) [inline services/manufacturerProdRangeService.js]: this try catch redundant and not needed.
check all try catches.

File: services/genericPipeService.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+// we get unitType as inches and sizes array with ids for reference.
+// 1. change add pipes to save array of ids in sizes.✅
+// 2. change get pipes to get the referenced data derived from the ids of sizes from master pipe.
+
+exports.getPagedGenericPipes = async (pageNo, docsPerPage) => {
+	try {
+		const totalDocs = await GenericPipe.countDocuments()
+		const genericPipeData = await GenericPipe.find()
+			.populate("sizes")
```
Comment (rohandhamapurkar) [inline services/genericPipeService.js]: need nosqlbosoter query for await GenericPipe.find().populate("sizes")

File: services/manufacturerProdRangeService.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	try {
+		// console.log(id, updateData)
+		return await ManufacturerProdRange.findByIdAndUpdate({ _id: id }, updateData)
+	} catch (error) {
+		// if passed data is already present inside the connection then throw error. but first check if error code is 11000
+		if (error.code !== 11000) throw error
+
+		error.statusCode = 400
+		error.message = "A manufacturer product range with the same data already exists."
+		throw error
```
Comment (rohandhamapurkar) [inline services/manufacturerProdRangeService.js]: errorObject function should be used here and whereever error.message is being set manually

**Review by rohandhamapurkar — CHANGES_REQUESTED** [review]

File: controllers/masterPipeController.js (hunk 1 above)
Comment (RajeevKumarSinha) [reply controllers/masterPipeController.js]: removed

File: services/manufacturerProdRangeService.js (hunk 7 above)
Comment (RajeevKumarSinha) [reply services/manufacturerProdRangeService.js]: done

File: services/genericPipeService.js (hunk 6 above)
Comment (RajeevKumarSinha) [reply services/genericPipeService.js]: db.genericpipes.aggregate(
    {
    $lookup:{
        from:"masterpipes",
        localField:"sizes",
        foreignField:"_id",
        as:"sizes"
    }
})

File: controllers/masterPipeController.js (hunk 2 above)
Comment (RajeevKumarSinha) [reply controllers/masterPipeController.js]: done

File: controllers/masterPipeController.js (hunk 3 above)
Comment (RajeevKumarSinha) [reply controllers/masterPipeController.js]: done

File: services/genericPipeService.js (hunk 4 above)
Comment (RajeevKumarSinha) [reply services/genericPipeService.js]: done

File: services/manufacturerProdRangeService.js (hunk 5 above)
Comment (RajeevKumarSinha) [reply services/manufacturerProdRangeService.js]: done

File: controllers/connectionTypeController.js [hunk 8]
```diff
@@ -41,7 +41,9 @@ exports.setConnectionType = async (req, res, next) => {
 			data: createdConnectionObj,
 		})
 	} catch (error) {
-		next(error)
+		if (error.code !== 11000) next(error)
```
Comment (rohandhamapurkar) [inline controllers/connectionTypeController.js:44]: if (error.code !== 11000)
this check should not be in the controller but in the service function instead.

Comment (rohandhamapurkar) [inline controllers/connectionTypeController.js:44]: if (error.code !== 11000)
this check should not be in the controller but in the service function instead.

its fine if you add try catch there

File: services/manufacturerService.js [hunk 9]
```diff
@@ -4,40 +4,25 @@ const { errObject } = require("../helpers/helper")
...
 exports.createManufacturer = async (manufacturerData) => {
-	try {
-		const manufacturer = await Manufacturer.create(manufacturerData)
-		return manufacturer
-	} catch (error) {
-		// console.log(error)
-		throw error
-	}
+	const manufacturer = await Manufacturer.create(manufacturerData)
+	return manufacturer
```
Comment (rohandhamapurkar) [inline services/manufacturerService.js:21]: cant we just not have 
`return await Manufacturer.create(manufacturerData)`
why do we need to initialize the `manufacturer` variable

File: controllers/connectionTypeController.js [hunk 10]
```diff
@@ -95,6 +97,8 @@ exports.updateConnectionType = async (req, res, next) => {
 		await connectionTypeService.patchConnectionType(id, updateDetail)
 		res.status(200).json({ status: "success", message: `Connection updated successfully.` })
 	} catch (error) {
-		next(error)
+		if (error.code !== 11000) next(error)
```
Comment (rohandhamapurkar) [inline controllers/connectionTypeController.js:100]: if (error.code !== 11000)
this check should not be in the controller but in the service function instead.

its fine if you add try catch there

File: controllers/genericPipeController.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+		const genericPipeReq = req.body
+		// console.log(genericPipeReq)
+		const createdGenericPipeObj = await genericPipeService.createGenericPipe(genericPipeReq)
+		res.status(201).json({
+			status: "success",
+			message: "New generic pipe created successfuly.",
+			data: createdGenericPipeObj,
+		})
+	} catch (error) {
+		if (error.code !== 11000) next(error)
```
Comment (rohandhamapurkar) [inline controllers/genericPipeController.js:31]: same here
if (error.code !== 11000)
this check should not be in the controller but in the service function instead.

its fine if you add try catch there

**Review by rohandhamapurkar — CHANGES_REQUESTED** [review]

**Comment by RajeevKumarSinha** [comment]
I have merged the branch as uday needed to get access to new API's

**Pull request merged**